stream/src/test/java/org/apache/apex/malhar/stream/sample/MyStreamTest.java

```diff
@@ -0,0 +1,121 @@
+import java.util.Map;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import org.apache.apex.malhar.stream.api.function.Function;
+import org.apache.apex.malhar.stream.api.impl.StreamFactory;
+
+/**
+ * A test class which test your own stream implementation build on default one
+ */
+public class MyStreamTest
+{
+  static Map<Object, Integer> expected = new HashMap<>();
+  static {
+    expected.put("newword1", 4);
+    expected.put("newword2", 8);
+    expected.put("newword3", 4);
+    expected.put("newword4", 4);
+    expected.put("newword5", 4);
+    expected.put("newword7", 4);
+    expected.put("newword9", 6);
+  }
+
+  @Test
+  public void testMethodChainWordcount() throws Exception
+  {
+    TupleCollector<Map<Object, Integer>> collector = new TupleCollector<>();
+    collector.id = "testMethodChainWordcount";
+    new MyStream<>(StreamFactory.fromFolder("./src/test/resources/data"))
+        .<String, MyStream<String>>flatMap(new Function.FlatMapFunction<String, String>()
+        {
+          @Override
+          public Iterable<String> f(String input)
+          {
+            return Arrays.asList(input.split(" "));
+          }
+        }).filterAndMap(new Function.MapFunction<String, String>()
+        {
+          @Override
+          public String f(String input)
+          {
+            return input.replace("word", "newword");
+          }
+        }, new Function.FilterFunction<String>()
+        {
+          @Override
+          public Boolean f(String input)
+          {
+            return input.startsWith("word");
+          }
+        }).countByKey()
+        .addOperator(collector, collector.inputPort, collector.outputPort).print().runEmbedded(5000);
+
+
+    List<Map<Object, Integer>> data = (List<Map<Object, Integer>>)TupleCollector.results.get("testMethodChainWordcount");
+    Assert.assertTrue(data.size() > 1);
+    Assert.assertEquals(expected, data.get(data.size() - 1));
+  }
+
+  @Test
+  public void testNonMethodChainWordcount() throws Exception
+  {
+    TupleCollector<Map<Object, Integer>> collector = new TupleCollector<>();
+    collector.id = "testNonMethodChainWordcount";
+    MyStream<String> mystream = new MyStream<>(StreamFactory
+        .fromFolder("./src/test/resources/data"))
+        .flatMap(new Function.FlatMapFunction<String, String>()
+        {
+          @Override
+          public Iterable<String> f(String input)
+          {
+            return Arrays.asList(input.split(" "));
+          }
+        });
+    mystream.filterAndMap(new Function.MapFunction<String, String>()
+    {
+      @Override
+      public String f(String input)
+      {
+        return input.replace("word", "newword");
+      }
+    }, new Function.FilterFunction<String>()
+    {
+      @Override
+      public Boolean f(String input)
+      {
+        return input.startsWith("word");
+      }
+    }).countByKey().addOperator(collector, collector.inputPort, collector.outputPort).print().runEmbedded(5000);
```

Comment:
The only existing API to shutdown a dag is throw ShutdownException but it doesn't guarantee the whole dag process all the data?